knimin/lib/data_access.py

```diff
@@ -3894,6 +3894,68 @@ def delete_shotgun_index_aliquot(self, shotgun_index_aliquot_id):
             TRN.add(sql, [shotgun_index_aliquot_id])
             TRN.execute()
 
+    def generate_i5_i7_indexes(self, idx_tech, num_samples):
+        """Pick barcodes for shotgun
+
+        Parameters
+        ----------
+        idx_tech : str
+            The index technology we want to use
+        num_samples : int
+            The number barcodes to generate
+
+        Returns
+        -------
+        list of ints
+            A list of ints that represent the barcodes selected for the samples
+
+        Raises
+        ------
+        ValueError
+            If idx_tech doesn't exist
+            If the idx_tech doesn't have barcodes
+        """
+        with TRN:
+            sql = """SELECT shotgun_index_tech_id, name, dual_index,
+                        last_index_idx
+                     FROM pm.shotgun_index_tech WHERE name = %s"""
+            TRN.add(sql, [idx_tech])
+            idx_tech_vals = TRN.execute_fetchindex()
+
+            if not idx_tech_vals:
+                raise ValueError('Not a valid shotgun index '
+                                 'technology: %s' % idx_tech)
+            else:
+                # just taking the first element and converting it to dict
+                idx_tech_vals = dict([v for v in idx_tech_vals][0])
+
+            # selecting the dual section
+            sql = """SELECT shotgun_index_id
+                     FROM pm.shotgun_index
+                     WHERE shotgun_index_tech_id = %s
+                     ORDER BY shotgun_index_id"""
+            TRN.add(sql, [idx_tech_vals['shotgun_index_tech_id']])
+            indices = [r[0] for r in TRN.execute_fetchindex()]
+            if not indices:
+                raise ValueError("'%s' doesn't have any barcodes" % idx_tech)
+            len_indices = len(indices)
+
+            idx = idx_tech_vals['last_index_idx']
+            indices_generated = []
+            for i in range(num_samples):
+                indices_generated.append(indices[idx])
+                idx += 1
+
+                if len_indices <= idx:
+                    idx = 0
+
+            sql = """UPDATE pm.shotgun_index_tech
+                     SET last_index_idx = %s
+                     WHERE shotgun_index_tech_id = %s"""
+            TRN.add(sql, [idx, idx_tech_vals['shotgun_index_tech_id']])
+
+            return indices_generated
```

Comment:
This is cool -- so really by storing the each combo as a separate entry in the DB, we don't have to have separate code for single and dual index? 👍